### PR TITLE
Add `ValidationConfig` for evaluator inference

### PR DESCRIPTION
**Pull request by jpdunc23**
Adds `validation_config: ValidationConfig` to `InferenceEvaluatorConfig` in `fme.ace`. When configured, validation runs first and logs metrics to step 0 of the wandb run.

Changes:
- Adds generic `run_validation` helper than can be reused by other modules.

- [x] Tests added

### CODE REVIEW

**Comment by jpdunc23** [comment]
Example wandb run: https://wandb.ai/ai2cm/ace2-cm4-atmos-noLSM/runs/2rlir6tb

**Comment by jpdunc23** [comment]
I looked into logging the validation metrics in the last step rather than step 0, which would be helpful since Wandb considers last-step metrics as [log summary metrics](https://docs.wandb.ai/models/track/log/log-summary), but I couldn't quite come up with a clean way to do it. It might be worth looking into it again.

File: fme/ace/inference/evaluator.py [hunk 1]
```diff
@@ -127,6 +165,10 @@ class InferenceEvaluatorConfig:
             This should be used with caution, as it may allow the stepper to make
             scientifically invalid predictions, but it can allow running inference with
             incorrectly formatted or missing grid information.
+        validation: Optional configuration for running a one-step validation loop
```
Comment (oliverwm1) [inline fme/ace/inference/evaluator.py:209]: I don't think "validation" the right word here. A user could run this part of the eval on any dataset, including the training data of the model being evaluated. The more relevant distinction is that this will compute/log "one step" metrics instead of the autoregressive inference metrics otherwise computed by the inference evaluator.

Comment (jpdunc23) [reply fme/ace/inference/evaluator.py:209]: > A user could run this part of the eval on any dataset, including the training data of the model being evaluated.

The same can be said of the `validation_loader` and corresponding `validation_aggregator` entries in `TrainConfig`.

> The more relevant distinction is that this will compute/log "one step" metrics

True that most of the logs will be "one step" metrics, but this supports multiple forward steps and reports "val/mean/loss", so calling this `one_step_eval` isn't quite right. 

`validation` still feels most appropriate to me since this is configured in exactly the same way as validation on `TrainConfig` (though that could change, of course).

Comment (oliverwm1) [reply fme/ace/inference/evaluator.py:209]: > The same can be said of the validation_loader and corresponding validation_aggregator entries in TrainConfig.

The context of `TrainConfig` is different since `TrainConfig` is actually the thing that defines the train/validation split.

> True that most of the logs will be "one step" metrics, but this supports multiple forward steps and reports "val/mean/loss", so calling this one_step_eval isn't quite right.

I agree "one step" is not quite right, but there should be some phrase that gets at "short time horizon metrics computed across many different data samples" to distinguish from the autoregressive metrics which are computed just from a smaller number of initial conditions, but potentially accumulated over long rollouts.

File: fme/ace/inference/evaluator.py [hunk 2]
```diff
@@ -79,6 +88,35 @@ def resolve_variable_metadata(
...
+            to 1 for standard single-step validation.
+        aggregator: Configuration for the one-step validation aggregator.
+        loss: Loss function configuration used to compute the validation loss
+            metric. Defaults to MSE. Set this to match the training loss
+            configuration if you want ``val/mean/loss`` to be directly
+            comparable to training.
+    """
+
+    loader: DataLoaderConfig
+    n_forward_steps: int = 1
```
Comment (mcgibbon) [inline fme/ace/inference/evaluator.py]: Unfortunate that this can't be used to achieve validation metrics for stochastic-n-steps training or for n_ensemble=2.

File: fme/ace/inference/evaluator.py (hunk 1 above)
Comment (mcgibbon) [reply fme/ace/inference/evaluator.py:209]: Without significantly refactoring the training code/config (which we could do, but I'm not sure is worth our time), it feels like calling this "validation" is the fastest/most obvious way to tell users what metrics these are computing. It's pretty valuable that this is called the same thing it's called in training, and the user story motivating this has been said as "we want to re-run then validation after the job is done or on different datasets". Calling this "short time horizon metrics" raises more questions than it answers, in my opinion, unless we also call it that somewhere in the training config / when we talk about training.

I do get that validation is "wrong" in the sense that this isn't a train/validation split, but I don't think a user will be confused about that, given they have to specify the dataset inside the config. To me "validation" in our discussions has been shorthand for "evaluation of an independent dataset on exactly the problem the model was trained on / run_on_batch"

File: fme/ace/inference/evaluator.py [hunk 3]
```diff
@@ -281,6 +324,38 @@ def run_evaluator_from_config(config: InferenceEvaluatorConfig):
...
+            n_forward_steps=config.validation.n_forward_steps
+        )
+        valid_data = get_gridded_data(
+            config.validation.loader,
+            requirements=data_requirements,
+            train=False,
+        )
+
+        logging.info("Building validation stepper and aggregator")
+        train_stepper_config = TrainStepperConfig(loss=config.validation.loss)
```
Comment (mcgibbon) [inline fme/ace/inference/evaluator.py]: Why limit to only being able to configure the loss? We could expose the entire TrainStepperConfig, or at least more of it. `train_n_forward_steps` and `n_ensemble` are both useful here.

File: fme/ace/inference/evaluator.py (hunk 2 above)
Comment (Arcomano1234) [reply fme/ace/inference/evaluator.py]: Yes I was thinking about making that as a follow on PR

File: fme/ace/inference/evaluator.py (hunk 3 above)
Comment (jpdunc23) [reply fme/ace/inference/evaluator.py]: OK, I think we can probably include that in this PR.

**Comment by jpdunc23** [comment]
Have a run that does stochastic ACE validation with `TimeLengthProbabilities`: https://wandb.ai/ai2cm/ace2-cm4-atmos-noLSM/runs/8ours1g1

A couple of unexpected things:
- ~~There was apparently no `val/mean/loss` logged. I think I just forgot to manually grab it from the `TrainOutput` and add it to the logs.~~ Nevermind, it's there but when I looked wandb wasn't showing it for some reason.
- The [logs](https://beaker.org/orgs/ai2/workspaces/climate-titan/work/01KKCZDS815B0WPJ4CTWBQTQBJ/logs?jobId=01KKCZDSRKCBDCY16TNKA3PC9S) this time for some reason don't include any entries like `Inference: processing output from window 207 of 1166.` This wasn't an issue with the [other run](https://beaker.org/orgs/ai2/workspaces/ace/work/01KJRCA1F150WAGAXD6HVA241Z/logs?jobId=01KJRCA1RMYS9D06QYSX1W4S98).

**Comment by Arcomano1234** [comment]
> * ~There was apparently no `val/mean/loss` logged. I think I just forgot to manually grab it from the `TrainOutput` and add it to the logs.~ Nevermind, it's there but when I looked wandb wasn't showing it for some reason.
> * The [logs](https://beaker.org/orgs/ai2/workspaces/climate-titan/work/01KKCZDS815B0WPJ4CTWBQTQBJ/logs?jobId=01KKCZDSRKCBDCY16TNKA3PC9S) this time for some reason don't include any entries like `Inference: processing output from window 207 of 1166.` This wasn't an issue with the [other run](https://beaker.org/orgs/ai2/workspaces/ace/work/01KJRCA1F150WAGAXD6HVA241Z/logs?jobId=01KJRCA1RMYS9D06QYSX1W4S98).

I wonder if the first issue is some how related to #693 

That second issue is interesting. I have not seen that before not sure why this would be related to a stochastic model with `TimeLengthProbabilities `

File: fme/ace/inference/evaluator.py (hunk 2 above)
Comment (mcgibbon) [reply fme/ace/inference/evaluator.py]: Whatever works for you.

File: fme/ace/inference/evaluator.py [hunk 4]
```diff
@@ -81,6 +93,78 @@ def resolve_variable_metadata(
...
+            to 1 for standard single-step validation.
+        aggregator: Configuration for the one-step validation aggregator.
+        stepper_training: Training-specific configuration including loss, ensemble
+            settings, and forward step scheduling. Set this to match the training
+            configuration if you want ``val/mean/loss`` to be directly comparable.
+
+    """
+
+    loader: DataLoaderConfig
+    n_forward_steps: int | None = None
```
Comment (mcgibbon) [inline fme/ace/inference/evaluator.py]: Question: Why include n_forward_steps if we already have `stepper_training.n_forward_steps`? You don't need to maintain backwards compatibility here, and it would simplify the code to use just one.

File: fme/ace/inference/evaluator.py [hunk 5]
```diff
@@ -286,6 +375,38 @@ def run_evaluator_from_config(config: InferenceEvaluatorConfig):
         stepper_all_names=stepper_config.all_names,
     )
     dataset_info = data.dataset_info.update_variable_metadata(variable_metadata)
+
+    if config.validation is not None:
+        timer.stop()
```
Comment (mcgibbon) [inline fme/ace/inference/evaluator.py:373]: Question: Why does this not count towards initialization?

File: fme/core/generics/validation.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        compute_derived_variables: Whether to compute derived variables.
+
+    Returns:
+        Dictionary of validation metrics (keys prefixed by the label).
+    """
+    if record_logs is None:
+        record_logs = _get_record_to_wandb()
+
+    timer = GlobalTimer.get_instance()
+    timer.start("validation")
```
Comment (mcgibbon) [inline fme/core/generics/validation.py]: Suggestion (optional): Use a `timer.context("validation"):` block to make it clearer what's timed and that the timer gets stopped appropriately.

File: fme/ace/inference/evaluator.py (hunk 5 above)
Comment (mcgibbon) [reply fme/ace/inference/evaluator.py:373]: Maybe do this in a follow on PR to get this one merged without needing another re-review, but the timer.stop and restart is reflecting the code here being a bit spaghetti - we're doing initialization, then taking a pause to do validation, then resuming initialization. I see it's because some local symbols like `aggregator` are re-used, which is a symptom of too much logic going on here and suggests we could use some helper methods.

If you can refactor this in a way that keeps initialization together such that we could use one `timer.context("initialization"):` instead of these stop-starts, it would be a bit easier to understand. Or using two `timer.context("initialization"):` blocks (and maybe one `timer.context("validation"):` block) would make the timer arithmetic here a little easier to follow.

File: fme/ace/inference/evaluator.py (hunk 4 above)
Comment (Arcomano1234) [reply fme/ace/inference/evaluator.py]: Wouldn't this allow you to have a validation `n_forward_steps` be a value different than what was used during training? I could see this being useful for weather forecasts where we want to know the day 3 skill or something.

File: fme/ace/inference/evaluator.py [hunk 7]
```diff
@@ -81,6 +93,78 @@ def resolve_variable_metadata(
...
+        """
+        train_n = self.stepper_training.train_n_forward_steps
+        if train_n is not None:
+            if isinstance(train_n, int):
+                return train_n
+            assert isinstance(train_n, TimeLengthProbabilities)  # already validated
+            return train_n.max_n_forward_steps
+        if self.n_forward_steps is not None:
+            return self.n_forward_steps
+        return 1
```
Comment (Arcomano1234) [inline fme/ace/inference/evaluator.py]: Optional Suggestion/Question: Do we want to have it log in the case where either `n_forward_steps` nor `stepper_training.train_n_forward_steps`  are configured that its using `n_forward_steps` of 1 ?

File: fme/ace/inference/evaluator.py (hunk 2 above)
Comment (jpdunc23) [reply fme/ace/inference/evaluator.py]: This should work in this PR now.

File: fme/ace/inference/evaluator.py (hunk 4 above)
Comment (jpdunc23) [reply fme/ace/inference/evaluator.py]: > Wouldn't this allow you to have a validation n_forward_steps be a value different than what was used during training?

You should be able to do this regardless with the `ValidationConfig.stepper_training.n_forward_steps`, so I agree that `ValidationConfig.n_forward_steps` is redundant here. Will remove.

File: fme/ace/inference/evaluator.py (hunk 5 above)
Comment (jpdunc23) [reply fme/ace/inference/evaluator.py:373]: OK, planning to handle via follow-on PRs.
